File: wrappers/javascript/shared/src/crypto/CryptoBox.ts
import type { Key } from './Key'

import { ariesAskar } from '../ariesAskar'

export class CryptoBox {
  public static randomNonce() {
    return ariesAskar.keyCryptoBoxRandomNonce()
  }

  public static cryptoBox({
    recipientKey,
    senderKey,
    message,
    nonce,
  }: {
    recipientKey: Key
    senderKey: Key
    message: Uint8Array
    nonce: Uint8Array
  }) {
    return ariesAskar.keyCryptoBox({ nonce, message, senderKey, recipientKey })
  }

  public static open({
    recipientKey,
    senderKey,
    message,
    nonce,
  }: {
    recipientKey: Key
    senderKey: Key
    message: Uint8Array
    nonce: Uint8Array
  }) {
    return ariesAskar.keyCryptoBoxOpen({ nonce, message, senderKey, recipientKey })
  }

  public static seal({ recipientKey, message }: { recipientKey: Key; message: Uint8Array }) {
    return ariesAskar.keyCryptoBoxSeal({ message, localKeyHandle: recipientKey.handle })
  }

  public static sealOpen({ recipientKey, ciphertext }: { recipientKey: Key; ciphertext: Uint8Array }) {
    return ariesAskar.keyCryptoBoxSealOpen({ ciphertext, localKeyHandle: recipientKey.handle })
  }
}

File: wrappers/kotlin/src/commonMain/kotlin/crypto/CryptoBox.kt
package askar.crypto

import askar.Askar

class CryptoBox {

    companion object {
        fun randomNonce(): ByteArray {
            return Askar.cryptoBox.keyCryptoBoxRandomNonce()
        }

        fun cryptoBox(recipientKey: Key, senderKey: Key, message: String, nonce: ByteArray): ByteArray {
            return Askar.cryptoBox.keyCryptoBox(recipientKey.handle(), senderKey.handle(), message, nonce)
        }

        fun cryptoBox(recipientKey: Key, senderKey: Key, message: ByteArray, nonce: ByteArray): ByteArray {
            return Askar.cryptoBox.keyCryptoBox(recipientKey.handle(), senderKey.handle(), message, nonce)
        }

        fun open(recipientKey: Key, senderKey: Key, message: ByteArray, nonce: ByteArray): ByteArray {
            return Askar.cryptoBox.cryptoBoxOpen(recipientKey.handle(), senderKey.handle(), message, nonce)
        }

        fun seal(recipientKey: Key, message: String): ByteArray {
            return Askar.cryptoBox.cryptoBoxSeal(recipientKey.handle(), message)
        }

        fun seal(recipientKey: Key, message: ByteArray): ByteArray {
            return Askar.cryptoBox.cryptoBoxSeal(recipientKey.handle(), message)
        }

        fun sealOpen(recipientKey: Key, cipherText: ByteArray): ByteArray {
            return Askar.cryptoBox.cryptoBoxSealOpen(recipientKey.handle(), cipherText)
        }

    }
}